### PR TITLE
Implement job scheduler infra

### LINKED ISSUE #4780: Implement job scheduler logic
Following #4693, implement the infrastructure of the job scheduling in Geti Tune.

_Note: the implementation of the training job is out of scope for this ticket._

### PR DESCRIPTION
**Pull request by itallix**
### Summary

Implements an asynchronous job controller system with process-based execution for job orchestration including training. This provides the foundation for managing concurrent ML jobs with proper resource management and event-driven lifecycle control. 

This is the adaptation of this [PR](https://github.com/itallix/jobs_demo/pull/1) for Geti Tune where the main goal is to separate job execution from lightweight async orchestration.

### How to test

**1. Submit a Test Job**

Use the curl command to submit a fake job:

```bash
curl -X POST -H "Content-Type: application/json" \
  localhost:7860/api/jobs \
  -d '{
    "job_type": "train", 
    "project_id": "20482f0c-a99e-4259-be5b-8edaee5f9045", 
    "parameters": { 
      "model_architecture_id": "YOLO_X" 
    }
  }'
```

This should return a response with a `job_id` that you'll use for monitoring.

**2. Monitor Job Progress with SSE**

Once you have the `job_id` from the submission response, monitor the job status:

```bash
curl -N localhost:7860/api/jobs/{job_id}/status
```

Replace `{job_id}` with the actual job ID. The `-N` flag disables buffering for real-time streaming.
You should see events like:
- `Started()` - Job execution begins
- `Progress(value=X)` - Progress updates (0-100%)
- `Done()` - Job completes successfully

**3. Cancel a Running Job**

To test cancellation, submit a job and then cancel it while it's running:

```bash
curl -X POST localhost:7860/api/jobs/{job_id}:cancel
```

You should see a `Cancelling()` and/or `Cancelled()` event in the SSE stream.

**4. List All Jobs**

To see all jobs that have been submitted during the current server session:

```bash
curl localhost:7860/api/jobs
```

This returns an array of all jobs with their current status, creation time, and other details.

### Checklist

- [x] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: application/backend/app/core/jobs/models/job.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    CANCELLED = 12
+
+
+def now_utc_ts() -> float:
+    """Get the current UTC timestamp as a float."""
+    return datetime.now(tz=UTC).timestamp()
+
+
+class Job(BaseModel):
+    id: UUID
```
Comment (leoll2) [inline application/backend/app/core/jobs/models/job.py]: Could be `BaseIDModel` instead of `BaseModel`?

File: application/backend/app/core/jobs/models/job.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    submitted_at: float
+    started_at: float | None = None
+    updated_at: float = now_utc_ts()
+    message: str | None = None
+    progress: float = 0.0  # percentage of completion
+    result: dict | None = None  # result of the job, if completed
+    error: str | None = None
+
+    @staticmethod
+    def new() -> "Job":
```
Comment (leoll2) [inline application/backend/app/core/jobs/models/job.py]: This method `Job.new()` isn't very pythonic; initialization with default values is typically done via the constructor.

```python
class Job(BaseModel):
    def __init__(
        self,
        id: UUID | None = None,
        status: JobStatus = JobStatus.PENDING,
        submitted_at: float | None,
        ...
    ) -> None:
        self._id = id or uuid4()
        self.submitted_at = submitted_at or now_utc_ts()
        ...
```

File: application/backend/app/core/jobs/models/job.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        )
+
+    def start(self) -> None:
+        self.status = JobStatus.RUNNING
+        self.started_at = now_utc_ts()
+        self.updated_at = now_utc_ts()
+
+    def advance(self, percent: float) -> None:
+        self.progress = max(0.0, min(100.0, percent))
+        self.updated_at = now_utc_ts()
```
Comment (leoll2) [inline application/backend/app/core/jobs/models/job.py]: Can it be extended with an optional `msg: str | None` parameter which, if provided, updates the job message? Or do suggest a different mechanisms for the jobs to output a custom message (e.g. "Downloading training weights...")?

File: application/backend/app/core/trainers/dummy.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+import logging
+import time
+
+from .base import Trainer, TrainerContext
+
+logger = logging.getLogger(__name__)
+
+
+class DummyTrainer(Trainer):
```
Comment (leoll2) [inline application/backend/app/core/trainers/dummy.py]: I think we don't need this `DummyTrainer`; you can move the dummy logic to `OTXTrainer`.

File: application/backend/app/core/jobs/models/events.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+@dataclass(frozen=True, slots=True)
+class Started(ExecutionEvent):
+    pass
+
+
+@dataclass(frozen=True, slots=True)
+class Progress(ExecutionEvent):
+    value: float
```
Comment (leoll2) [inline application/backend/app/core/jobs/models/events.py]: Maybe rename to `percentage` or add a docstring to clarify the range of this value ([0-100] vs [0-1])

File: application/backend/app/core/jobs/models/job.py (hunk 1 above)
Comment (itallix) [reply application/backend/app/core/jobs/models/job.py]: I will change to `dataclass`. We don't have any validation at this level where pydantic can be useful.

File: application/backend/app/core/jobs/models/job.py (hunk 3 above)
Comment (itallix) [reply application/backend/app/core/jobs/models/job.py]: I will include `msg` for now. It can be extracted in a separate method in the future if needed.

File: application/backend/app/core/jobs/models/job.py (hunk 1 above)
Comment (itallix) [reply application/backend/app/core/jobs/models/job.py]: It turns out serialization/deserialization can still be useful.

I have to move `models/base.py` to `core/models.py`. It makes more sense if we plan to share those models not only with schema-specific classes.

File: application/backend/app/api/endpoints/jobs.py [hunk 6]
```diff
@@ -108,18 +120,33 @@ async def cancel_job(job_id: JobID) -> JobResponse:
...
+        job, result = job_queue.cancel(job_id)
+        match result:
+            case CancellationResult.NOT_FOUND:
+                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job not found")
+            case CancellationResult.IGNORE_CANCEL:
+                raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Job already completed or cancelled")
+            case CancellationResult.PENDING_CANCELLED | CancellationResult.RUNNING_CANCELLING:
+                if job:
+                    return JobView.of(job)
+                raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Job not found")
```
[the remark this answers is not in the document]
Comment (itallix) [reply application/backend/app/api/endpoints/jobs.py]: To pass mypy checks

File: application/backend/app/core/jobs/control_plane/process_run.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            kill_timeout: How long to wait after SIGKILL (seconds)
+        """
+        if self._proc is None:
+            return
+
+        # Request graceful shutdown
+        self._cancel.set()
+
+        # Try graceful shutdown
+        await asyncio.to_thread(self._proc.join, timeout=graceful_timeout)
```
Comment (leoll2) [inline application/backend/app/core/jobs/control_plane/process_run.py:100]:
```suggestion
        # Request graceful shutdown
        self._cancel.set()
        await asyncio.to_thread(self._proc.join, timeout=graceful_timeout)
```

**Review by leoll2 — DISMISSED** [review]

File: application/backend/tests/integration/core/jobs/control_plane/thread_run.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+import threading
+import time
+from collections.abc import Iterator
+from queue import Empty, Queue
+
+from app.core.jobs.models import Cancelled, Done, ExecutionEvent, Failed, Job, Progress, Started
+from app.core.run import ExecutionContext, RunnableFactory, Runner
+
+
+class ThreadRunner(Runner[Job, ExecutionEvent]):
```
Comment (leoll2) [inline application/backend/tests/integration/core/jobs/control_plane/thread_run.py]: What about moving this runner implementation to `core.jobs.control_plane`, next to the process runner? It would allow us to quickly experiment with thread-based jobs for development purposes. If you don't want to create tests for it, you can raise a warning that the class is "experimental".

File: application/backend/app/core/jobs/control_plane/queue.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+logger = logging.getLogger(__name__)
+
+
+class CancellationResult(StrEnum):
+    """Outcome of a job cancellation request."""
+
+    PENDING_CANCELLED = "pending_cancelled"
+    RUNNING_CANCELLING = "running_cancelling"
+    IGNORE_CANCEL = "ignore_cancel"
+    NOT_FOUND = "not_found"
```
Comment (leoll2) [inline application/backend/app/core/jobs/control_plane/queue.py:28]: Can you add a comment/dosctring to explain the meaning of these values?

File: application/backend/tests/integration/core/jobs/control_plane/thread_run.py (hunk 8 above)
Comment (itallix) [reply application/backend/tests/integration/core/jobs/control_plane/thread_run.py]: Good idea. I've added a warning with a proper message.

File: application/backend/app/core/jobs/control_plane/queue.py (hunk 9 above)
Comment (itallix) [reply application/backend/app/core/jobs/control_plane/queue.py:28]: Added.

**Review by leoll2 — APPROVED** [review]
